internal/xdscache/v3/endpointslicetranslator.go

```diff
@@ -0,0 +1,458 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	envoy_endpoint_v3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	resource "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/projectcontour/contour/internal/contour"
+	"github.com/projectcontour/contour/internal/dag"
+	envoy_v3 "github.com/projectcontour/contour/internal/envoy/v3"
+	"github.com/projectcontour/contour/internal/k8s"
+	"github.com/projectcontour/contour/internal/protobuf"
+	"github.com/projectcontour/contour/internal/sorter"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+	v1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+// RecalculateEndpointsSlice generates a slice of LoadBalancingEndpoint
+// resources by matching the given service port to the given discoveryv1.EndpointSlice.
+// endpointSliceMap may be nil, in which case, the result is also nil.
+func (c *EndpointSliceCache) RecalculateEndpoints(port, healthPort v1.ServicePort, endpointSliceMap map[string]*discoveryv1.EndpointSlice) []*LoadBalancingEndpoint {
+	var lb []*LoadBalancingEndpoint
+	uniqueEndpoints := make(map[string]struct{}, 0)
+	var healthCheckPort int32
+
+	for _, enpointSlice := range endpointSliceMap {
+		sort.Slice(enpointSlice.Endpoints, func(i, j int) bool {
+			return enpointSlice.Endpoints[i].Addresses[0] < enpointSlice.Endpoints[j].Addresses[0]
+		})
+
+		for _, endpoint := range enpointSlice.Endpoints {
+			// Skip if the endpointSlice is not marked as ready.
+			if endpoint.Conditions.Ready != nil && !*endpoint.Conditions.Ready {
+				continue
+			}
+
+			// Range over each port. We want the resultant endpoints to be a
+			// a cartesian product MxN where M are the endpoints and N are the ports.
+			for _, p := range enpointSlice.Ports {
+				// Nil check for the port.
+				if p.Port == nil {
+					continue
+				}
+
+				if p.Protocol == nil || (healthPort.Protocol != *p.Protocol || port.Protocol != *p.Protocol) && *p.Protocol != v1.ProtocolTCP {
```

Comment:
Still digesting the logic in this line, I feel like the `&&` should be an `||` (I know this matches what was in EndpointsTranslator though). We can address separately if it should be changed, since it would need to change in both if so.

Reply:
yeah looking at it again, it feels like it should be an `||` because we only want to support TCP, we should short-circuit it if the protocol is anything else .

Reply:
I havent changed this yet, but happy to change this here and create an issue for `endpoints`